mpi.h

```diff
@@ -577,10 +577,18 @@ typedef void (MPI_T_event_free_cb_function)(MPI_T_event_registration event_regis
 typedef void (MPI_T_event_dropped_cb_function)(MPI_Count count, MPI_T_event_registration event_registration, int source_index, MPI_T_cb_safety cb_safety, void *user_data);
 
 /* MPI global variables */
-extern MPI_Fint* MPI_F_STATUS_IGNORE;
-extern MPI_Fint* MPI_F_STATUSES_IGNORE;
-extern MPI_F08_status* MPI_F08_STATUS_IGNORE;
-extern MPI_F08_status* MPI_F08_STATUSES_IGNORE;
+extern void*                           MPI_F_BOTTOM;
+extern void*                           MPI_F_IN_PLACE;
+extern void*                           MPI_F_BUFFER_AUTOMATIC;
+extern char**                          MPI_F_ARGV_NULL;
+extern char***                         MPI_F_ARGVS_NULL;
+extern MPI_Fint*                       MPI_F_ERRCODES_IGNORE;
+extern MPI_Fint*                       MPI_F_STATUS_IGNORE;
+extern MPI_Fint*                       MPI_F_STATUSES_IGNORE;
+extern MPI_F08_status*                 MPI_F08_STATUS_IGNORE;
+extern MPI_F08_status*                 MPI_F08_STATUSES_IGNORE;
+extern MPI_Fint*                       MPI_F_UNWEIGHTED;
+extern MPI_Fint*                       MPI_F_WEIGHTS_EMPTY;
```

Comment:
Why do we need `MPI_F_BOTTOM` etc in `mpi.h`? They aren't in the MPI standard, are they?

Reply:
Indeed, they are not yet part of the MPI standard yet. 

This is a new thing that  Jeff and I were discussing adding to help tools intercepting Fortran calls in C. We had yet another more ambitious objective, that is, use these variables to eventually define the Fortran-side things (via `BIND(C)`) to these C symbols in the C library. However, at this point, portability is not clear, at least in macOS (related to the now broken `-Wl,-commons,use_dylibs` flag).

Even if the ambitious objective is not clear to be achievable, the secondary objective of exposing these Fortran-side constants to C world may prove useful to users, see last sentence here: https://github.com/mpi-forum/mpi-issues/issues/837#issuecomment-1971139198.

Of course, as part of the MPI 4.2 update, MPICH will eventually require some minimal changes to assign to these global variables during MPI initialization. But I guess there is no rush for the time being.

Reply:
I am concerned that we are assuming a C-ABI for all MPI implementation. Let's say eventually we want to implement MPI entirely in Rust, or a future Fortran,  or a future system language, the assumption of a C-ABI may make such implementation very constrained. I view MPI-ABI as an MPI derivative, thus am concerned that it is starting back impact the MPI for other bindings.

Reply:
By the way, I am treating this repository, especially `mpi.h`, as upstream truth. If non-standard experiments or extensions start to appear in it, I have to update my view.

Reply:
I'm not following your concern in full. Nonetheless, please note that the MPI standard already throwed in the face of implementations the requirement for exposing `MPI_F[08]_STATUS[ES]_IGNORE`. All what we are doing here is extending the exposure to other link constants that cannot be implemented in Fortran as compile-time constants.

> the assumption of a C-ABI may make such implementation very constrained.

What specific constraint are you worried about from my last commit here? As I said, we already have to, for better or worse, implement `MPI_F[08]_STATUS[ES]_IGNORE` in C for Fortran to initialize these variables. This whole mess is not actually about C ABI, but rather Fortran mess leaking out of Fortran to C land.

Reply:
> So yes, I prefer all non-standard symbols either in comments or in `MPIX_` prefix.

Fair enough, this was definitely an oversight of me, thanks for catching it up.

Reply:
> I'm not following your concern in full. Nonetheless, please note that the MPI standard already throwed in the face of implementations the requirement for exposing `MPI_F[08]_STATUS[ES]_IGNORE`. All what we are doing here is extending the exposure to other link constants that cannot be implemented in Fortran as compile-time constants.
> 
> > the assumption of a C-ABI may make such implementation very constrained.
> 
> What specific constraint are you worried about from my last commit here? As I said, we already have to, for better or worse, implement `MPI_F[08]_STATUS[ES]_IGNORE` in C for Fortran to initialize these variables. This whole mess is not actually about C ABI, but rather Fortran mess leaking out of Fortran to C land.

A precedent does not mean it is the right choice. MPI has made plenty of non-optimum choices before. Maybe we can agree it's a mess. The least we can do is resist to extend the mess.

The Fortran/C interop is an after thought. I agree it is practically useful and important for the time being, but I think it should stay in a side document.

Reply:
> By the way, I am treating this repository, especially `mpi.h`, as upstream truth. If non-standard experiments or extensions start to appear in it, I have to update my view.

OK, look.. .I'm going to do the following. I'll split this PR in two. The new MPIX_F_ things will be addressed in an upcoming separate PR. This way, we can merge this one and the MPICH one.

Reply:
Done. Could you please relax the MPIX_ prefix rule for  `MPI_ABI_[SUB]VERSION`?

Reply:
> Done. Could you please relax the MPIX_ prefix rule for `MPI_ABI_[SUB]VERSION`?

I guess that's fine. We can view `MPI_ABI_` as its own name space.